mautic/mautic
# Feature builder secitons

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | Y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/163
| Related developer documentation PR URL | https://github.com/mautic/developer-documentation/pull/51
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/3439, https://github.com/mautic/mautic/issues/3148, possibly https://github.com/mautic/mautic/issues/3417
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR improves the Email/Page Builder sections. They can be now moved up or down, deleted and 3 new section types can be added. With 1,2 and 3 column layout.

![mautic-builder-sections](https://cloud.githubusercontent.com/assets/1235442/23134837/8970b30e-f796-11e6-891f-ae258828ce91.png)

+ The Medium and Large button sizes are bigger.
+ Fixed JS errror `Editor should be initialized before calling the destroy method.` when closing the builder.
+ Image slot editor fixed, aligning works.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. This PR also fixes an issue where sometimes the slots couldn't be moved or added new ones. They could be only edited. When I checked the content, the `data-*` which didn't have any value HTML attributes were removed. I added a value to them so it should fix it. I didn't have an issue with it after. I don't know how to reproduce it.
2. The dev tools console throws errors and warnings on different actions - is fixed in this PR.
3. In the code mode try to save this content:
```

```
The `<o:*>` and `<xml>` tags will get stripped. They won't after this PR. The unit test added to ensure that for the future updates. Feel free to run unit tests.

#### Steps to test this PR:
1. Apply this PR.
2. Clear the cache.
3. Rebuild Assets with `app/console mautic:assets:generate` command.
4. Try to edit/create email/page with different themes. The old functionality should work as before and you should be able to order/delete/create sections.
5. Test that the image slot editor aligns the image to left, right or center.
